database/mysql.go

```diff
@@ -113,12 +114,27 @@ type TestMySQLDB struct {
 func (r *TestMySQLDB) Close() error {
 	r.shutdown()
 
+	// ASK: not sure what do we check in the following.
+	// We don't always close sql.DB directly but call
+	// this Close method.
```

Comment:
I'd like to maintain this check. It's saved myself from a bug where i had an incorrect transaction setup in my repository which led to a connection left open (and eventually connection pool exhaustion).

Reply:
This check runs at the end of a `func TestXXX(t *testing.T)` method so the underlying connection(s) should already be closed.

Reply:
The strange thing about this code is:
we call TestMySQLDB.Close() which checks if we closed DB and if yes then it will close DB :D 

```
# we want to close DB
func (r *TestMySQLDB) Close()
...
# did we closed DB?
if conns := r.DB.Stats().OpenConnections; conns != 0 {
  # no, we didn't close DB
  # panic
 ...
}

# close DB
err := r.DB.Close()
```

maybe I don't understand the flow or miss something here...

I may guess that some code, having access to TestMySQLDB.DB may close it before calling TestMySQLDB.Close(). But at least in one case I was not able to do this...

Reply:
hmm we've been calling `db.Close()` but it doesn't look like it's necessary. The mention in code comments in `database/sql`:
```
// Close closes the database and prevents new queries from starting.
// Close then waits for all queries that have started processing on the server
// to finish.
//
// It is rare to Close a DB, as the DB handle is meant to be
// long-lived and shared between many goroutines.
func (db *DB) Close() error {
```

and I've read the connection pool handles reuse and closing.

Reply:
We should `Close()` the Conn and DB objects in our tests. Yes, in production code it's rare to close an `*sql.DB`, but in tests it's common as cleanup.

Reply:
Guys, the question here is not close or not to close.

But the code itself. It does not allow to call DB close if it's not closed 🤷‍♂️

I just want to remove the code that checks if everything was closed as I don't understand it and it seems counter-intuitive to me.

Reply:
This check helped me avoid bugs. It's a great check to include. I don't understand why we'd remove it.

Reply:
Here's the example of this check on `.OpenConnections` helped me diagnose a connection leak. Let's keep this code and prevent bugs in the future. 

https://github.com/moov-io/paygate/pull/344

Reply:
From sql.go mentioned by @vxio :

// Close closes the **database** and prevents new queries from starting.
// Close then waits for all queries that have started processing on the server
// to finish.
//
// It is rare to Close a DB, as the DB handle is meant to be
// long-lived and shared between many goroutines.
func (db *DB) Close() error {
...
}

As I understand it:

* we connect to a database (we do this once in an app)
* DB driver maintains pool of connections (automatically, based on set max con, idle con, etc.)
* if there is no free connection for sql query it will automatically open a new connection (invisibly to user) and use it for the query and then keep it in the pool of connections
* we do not close connections from pool as we don't have control over them
* We close (global) connection to the database which will close all opened connections inside the pool after queries finish / or time out

I did some experiments to double check my understanding and show that our `Close` method should be changed as it does not work it supposed to work. It worked before because in most cased in our tests we close `sql.DB` before we call `TestMySQLDB.Close` method (and goal of `TestMySQLDB.Close` is exactly to close DB).

Here is first experiment:

```go
func TestMySQLExperiment1(t *testing.T) {
	db := CreateTestMySQLDB(t)
	db.Close()
}

# fails
# --- FAIL: TestMySQLExperiment1 (18.14s)
#panic: found 1 open MySQL connections [recovered]
#        panic: found 1 open MySQL connections
```

I would expect it to work.

second experiment to show that we do not control (open or close) connections in the pool:

```go
func TestMySQLExperiment2(t *testing.T) {
	db := CreateTestMySQLDB(t)
	defer db.Close()

	ticker := time.NewTicker(100 * time.Millisecond)
	done := make(chan bool)

	go func() {
		for {
			select {
			case <-done:
				return
			case <-ticker.C:
				fmt.Printf("(1) open connections: %d\n", db.DB.Stats().OpenConnections)
			}
		}
	}()

	performQuery := func() {
		for {
			rows, _ := db.DB.Query("select * from tests")
			rows.Close()
			time.Sleep(50 * time.Millisecond)
		}
	}

	go performQuery()
	go performQuery()
	go performQuery()

	time.Sleep(2 * time.Second)
	ticker.Stop()
	done <- true
}

# output
# go test -run TestMySQLExperiment2
# (1) open connections: 1
# (1) open connections: 1
# (1) open connections: 1
# (1) open connections: 1
# (1) open connections: 4
# (1) open connections: 1
# (1) open connections: 4
# (1) open connections: 1
# (1) open connections: 1
# (1) open connections: 1
# (1) open connections: 1
# (1) open connections: 1
# (1) open connections: 4
# (1) open connections: 1
# (1) open connections: 1
# (1) open connections: 1
# (1) open connections: 1
# (1) open connections: 4
# (1) open connections: 1
# (1) open connections: 4
# --- FAIL: TestMySQLExperiment2 (20.20s)
# panic: found 4 open MySQL connections [recovered]
#         panic: found 4 open MySQL connections
```

From 2nd experiment we can see that we opened only one connection, but somehow got 4 opened and we also see that there is no way for us to close... 3 of them.

Finally,
* we need to remove commented lines from our `Close` method and
* just close DB together with docker container.

Reply:
That's interesting.. It seems odd to me we can't close the `*sql.DB` in tests, but that appears to be from the ordering of `defer` and `t.Cleanup` calls. I think `t.Cleanup` is called before before our `defer` calls if we were to get the call stack.

Reply:
Experiment 3 (just create test mysql db and see how many open connections we have from the start -> we expect to have 0):

```go
func TestMySQLExperiment3(t *testing.T) {
	db := CreateTestMySQLDB(t)
        defer db.Close()
	ticker := time.NewTicker(100 * time.Millisecond)
	done := make(chan bool)
	go func() {
		for {
			select {
			case <-done:
				return
			case <-ticker.C:
				fmt.Printf("open connections: %d\n", db.DB.Stats().OpenConnections)
			}
		}
	}()
	time.Sleep(2 * time.Second)
	ticker.Stop()
	done <- true
}
```
the result:

![image](https://user-images.githubusercontent.com/19233/96141820-ae626500-0f01-11eb-968a-bf81c14b11ee.png)

Reply:
Our assumption that we somehow have 0 open connections is wrong. I don't know why. Maybe it's because of db config we have right now (max open connections, max idle connections, etc.).

Reply:
I found one place that increases opened connections from 0 to 1. It's `err = RunMigrations(logger, db, *config)` which opens connection. Now the question, why it does not close it.

Here is the discussion on this topic https://github.com/golang-migrate/migrate/issues/97
https://github.com/golang-migrate/migrate/issues/97#issuecomment-464445111

@adamdecaf the issue is with migrator that opens one more connection that should be closed (released). As it accepts "db" that created for the app we can't call "Close()" on it. Solution is to create to separate db instance for migrations and close it.

Reply:
Solved by using [separate DB connection](https://github.com/moov-io/base/pull/96/files#diff-7c9b482ada7e6e60f6c5c4593eaed1bb967e0813d28b63e278a9649046636395R19) for migrator.